### ISSUE TITLE
Fix dependency vulnerability d3-color < 3.1.0

### ISSUE THREAD

**Issue by StanDeveloper** [recharts/recharts#3012]
- [ ] I have searched the [issues](https://github.com/recharts/recharts/issues) of this repository and believe that this is not a duplicate.

### What problem does this feature solve?
Updating the package will, fix the Redos vulnerability of the d3-color; https://github.com/advisories/GHSA-36jr-mh4h-2g58

### What does the proposed API look like?
No api needed

**Comment by timges** [comment]
Wanted to post the same. Imo this should have high priority as the vulnerability is of high severity. 👍🏼

**Comment by zavan** [comment]
I think #3009 solves this.

**Comment by timges** [comment]
Recharts downgraded the version's again with today's release, so i guess dependabot's pr is not sufficient, as the checks are failing anyways

**Comment by vitalii-bulyzhyn** [comment]
> Release 2.1.15 downgraded the version's again today. I guess dependabot's pr ( #3009 ) is not sufficient, as the checks are failing anyways

+ 1 for this, still the same problem

**Comment by nftnftnftgem** [comment]
+1

**Comment by emirefek** [comment]
Are there any fix for this? What do you people suggest?
My dependabot and npm audit really going crazy about it.

**Comment by mark-todd** [comment]
Dependabot also has a PR for this it turns out #3009

**Comment by G-Rath** [comment]
@mark-todd it's not that simple because v3 of `d3-color` is ESM based, meaning its not directly compatible with `recharts` - the upgrade you're suggesting has actually already been done and then [reverted](https://github.com/recharts/recharts/commit/a78ed04256a9497cfbc6645ff0522a413daf9f75) in v2.1.15 because it was breaking.

To resolve this properly one of three things needs to happen:
  1. `recharts` does a new major version that switches it to using v3 of `d3-color` (via the other `d3` dependencies)
    - `recharts` itself technically doesn't have to move to ESM, but it's still a breaking change because downstream consumers need to support ESM or transpile the modules
  2. someone forks the related modules, publishes them to npm, and switches recharts over to using that
    - this would be a lot of work, and just be an overall annoyance; but maybe someone out there is able to spare the time?
  3. `d3-color` releases a v2 with the fix
    - this would be the most optimal, though @mbostock has said [they won't be doing this](https://github.com/d3/d3-color/issues/108#issuecomment-1271705762); I have [asked them to reconsider](https://github.com/d3/d3-color/issues/108#issuecomment-1282822581) but I'm not super hopeful

**Comment by mark-todd** [comment]
Hi @G-Rath - thanks for the detailed response! Option 1 looks pretty good to me - is there any reason not to make a major release for this? If a major security vulnerability doesn't qualify as a reason for a new major release - what does?

**Comment by G-Rath** [comment]
@mark-todd because it would require downstream consumers to be able to upgrade to that new major version, and in this case that would require being able to use ESM which is not an option for a lot people - so it would be just pushing the problem further downstream.

**Comment by mark-todd** [comment]
Hmm true - based on this comment: https://github.com/d3/d3-color/issues/108#issuecomment-1275818206 ...it seems like the v2 with fix version already exists. It won't be on npm, but it could be imported to Recharts with the git package syntax I suppose?

**Comment by G-Rath** [comment]
No, that is just a PR someone has made to the `main` branch (which is v3) asking that it be ported to v2 since a dedicated branch for v2 doesn't exist so there's no way to make a PR for it.

Using git-syntax would technically resolve the problem but that means `npm` has to do more complex work when installing (as it needs to use `git` to clone the repo, and that also requires authentication even ifs public) and would no longer be covered by security advisories so you'd not be able to detect if you're impacted by any future advisories for the package.

It's also not just that which'd bre required as `recharts` is not depending on `d3-color` directly - it's being pulled in by at least `d3-interople` so you'd actually need to either get the v2 of that to switch to using a different dependency (which seems unlikely to happen since they're not interested in doing a v2 backport for `d3-color`, which'd be less work than backporting for `d3-interople`), or fork that package and have that as a git dependency or publish it as a fork to the npm registry.

At this point unless @mbostock changes their mind, option 2 (forking a bunch of the d3 modules and publishing them to npm) is probably going to be the "best" option, if someone has bandwidth to maintain that.

(Having said that an alternative worth exploring first is if `recharts` can move off any of the d3 modules - even if only some of them can be replaced, it's probably going to make things easier)

**Comment by gergokee** [comment]
OK, so are there no plans for this ? Should we switch to another chart solution?

**Comment by mark-todd** [comment]
Yeah I agree with @gergokee on this - I know there are backwards compatibility concerns with the ESM part, but for people trying to start using recharts (and a lot of existing developers), this will mean we have to find other charting libraries. Apart from being a real shame for recharts, this will require a lot of overhaul for the existing developers - it effectively makes all of recharts a deprecated library.

**Comment by OhJayDubya** [comment]
@mark-todd So I am guessing that fixes like #3009 and #3022 aren't going to work due to the ESM issue? Just currently weighing up my options on whether to move off of recharts or hope these fixes will stop my repo from getting flagged for the d3-color vulnerability.

**Comment by wmhartl** [comment]
Agree with the posters above, recharts should bump its major version as a breaking change and move forward here. That's a common tactic in our ecosystem. We'd also like to stick with recharts versus moving to another package.

**Comment by kevbrowndev** [comment]
What's the best way folks have found to work around this weekend while we wait for Recharts to be updated?  Is replacing it with another library or a doctored version of Recharts really the best way to go?

**Comment by cfry2** [comment]
If you're using a bundler like Webpack or Rollup as well as using yarn you can add to your package.json

`  "resolutions": {
    "d3-color": "3.1.0"
  }`

In order for this to play nice with Jest, add in your jest.config.js

`  transformIgnorePatterns: ['<rootDir>/node_modules/(?!(d3-interpolate|d3-color)/)'],`

This assume's you're using babel to transpile your code

This works for me.

**Comment by kopach** [comment]
and if you're lucky to use `pnpm` as package manager you can do like this for now:
in `package.json` add

```json
  "pnpm": {
    "overrides": {
      "d3-interpolate@2.0.1>d3-color": "~3.1.0"
    }
  }
```
docs: https://pnpm.io/package_json#pnpmoverrides

**Comment by norbulcz** [comment]
Hello, First of all I appreciate what everybody has done with this library, I think is great. I just want to ask if is there an intention to fix this issue? Would be great to know if we can rely on this or try a workaround.

**Comment by kozzza** [comment]
What's the best workaround for npm users? I'm trying to override the ``d3-interpolate`` dependency, but am getting this error when installing recharts: 

``npm ERR! Override for d3-interpolate@^3.0.1 conflicts with direct dependency``

Also can we get a timeline on the resolution of this dependency vulnerability? Thank you.

**Comment by G-Rath** [comment]
> Also can we get a timeline on the resolution of this dependency vulnerability?

@kozzza see my comments [here](https://github.com/recharts/recharts/issues/3012#issuecomment-1282835425) and [here](https://github.com/recharts/recharts/issues/3012#issuecomment-1284393411).

At this point it doesn't look like there will be a fix that is compatible with `recharts` as is it today since @mbostock has said they won't be doing a backport, and a significant amount of work would be required to address it at the `recharts` layer.

I think the "best" middleground (which would still require some work) would be for someone publish forks of the necessary parts of the d3 ecosystem from the pre-ESM versions so that the security patch can be applied, and then `recharts` can switch over to using those without requiring a breaking change.

**Comment by kozzza** [comment]
Understood, thanks for the insight @G-Rath.

**Comment by nfiacco** [comment]
@G-Rath is there any reason not to release a major version change for those of us who can use ESM in addition to your proposed fix for V2? Would love a version that is depending on the un-forked version of D3.

**Comment by G-Rath** [comment]
@nfiacco because someone would have to spend the time making that change (implementing, testing, documenting, releasing, etc), and then maintain two major versions (or otherwise gate off a huge number of devs that are not able to use ESM).

But ultimately it's up to the maintainers of this library to decide if they want to go down that path or not :)

**Comment by nyg** [comment]
I think I found a patched version of d3-color but v1 only: https://github.com/mpopv/d3-color-1-fix

Could not find anything for the v2.

**Comment by ckifer** [comment]
Hey all, sorry this issue hasn't gotten any response or action. Recharts has had no devs got quite some time. Will try to bring this issue up and see if a plan can be made to mitigate. The fact that they won't backport is quite a shame...

**Comment by ckifer** [comment]
@myasteiner sounds good to me as long as we're certain nothing breaks. 

It might not be merged right away but I'll do my best to help push it. I can see how blocking it is for people. 

Then, if we happen to need truly need this lib later we can bump to 3.0 (or at least that would be the idea. 

If you're interested in contributing [join the slack](https://github.com/recharts/recharts/discussions/2970#discussioncomment-4339416).

**Comment by mrrau** [comment]
So much trouble with so simple solution. Maybe the safest thing will be just return const array 

 `conset schemeCategory10 = [ "#1f77b4", "#ff7f0e", "#2ca02c", "#d62728", "#9467bd", "#8c564b", "#e377c2", "#7f7f7f", "#bcbd22", "#17becf" ];`

and it can be updated whenever https://github.com/d3/d3-scale-chromatic/blob/main/src/categorical/category10.js will change (probably never)

**Comment by nikolasrieble** [comment]
How would that help though? As we still use d3-scale in other places, we would still depend on d3-color, see `npm audit`

```
# Run  npm install d3-scale@4.0.2  to resolve 1 vulnerability
SEMVER WARNING: Recommended action is a potentially breaking change
┌───────────────┬──────────────────────────────────────────────────────────────┐
│ High          │ d3-color vulnerable to ReDoS                                 │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Package       │ d3-color                                                     │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Dependency of │ d3-scale                                                     │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Path          │ d3-scale > d3-interpolate > d3-color                         │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ More info     │ https://github.com/advisories/GHSA-36jr-mh4h-2g58            │
└───────────────┴──────────────────────────────────────────────────────────────┘

```

**Comment by ckifer** [comment]
Leaves me a bit at a loss - we can't easily remove the dependency on `d3-scale`. We might as well fork `d3-color` at that point. Could we try to support both ESM and commonjs?

**Comment by ckifer** [comment]
Forked and fixed v2 in https://github.com/ckifer/d3-color-2-fix for fun and [published](https://www.npmjs.com/package/d3-color-2-fix)

**Comment by ckifer** [comment]
https://formidable.com/blog/2022/victory-esm/

Looks like the solution from Victory is the way we have to go

**Comment by ckifer** [comment]
@myasteiner 1.4.1 still has the CVE in it no?

**Comment by ckifer** [comment]
I'm thinking of just using `victory-vendor` directly. It's open-source and vended as an npm package 🤔. It would solve all issues and be incredibly easy. Not sure how they'd feel that another charting library is using it, but it's vended.

**Comment by jmfrancois** [comment]
The downgrade happens because latest d3 ecosystem do not fit with current tools limitation. They publish package using esm and it makes jest, nextjs and other tools more hard to configure to work around that issue.

You can look at many opened issues:
#2991
#2998
#2915

good luck

**Comment by wuxinange** [comment]
this is a long thread, could anyone share the final solution or mitigation for this?

**Comment by ckifer** [comment]
@wuxinange 

recharts will replace d3 libraries with `victory-vendor` versions of d3 libraries so we can continue to vend commonjs and not require ESM. This should fix the vulnerabilities for everyone. Eventually if we have more need to vend ESM only then we will revert back to d3 libraries at v3 and bump our major version.

Recharts hasn't had active devs in a long time and a few of us are attempting to make the developer experience/tests/process better before we make any huge changes. Sorry this vulnerability has been here for so long and thanks everyone for your continued patience.

**Comment by ckifer** [comment]
Okay I have two branches that prove a non-breaking fix.

https://github.com/ckifer/recharts/tree/test/d3FixBreakBuild - this branch breaks the build with ESM errors since we now run jest tests in Recharts. This catches the issue which required 2.1.13 being reverted in the build step.

https://github.com/ckifer/recharts/tree/patch/d3-vulnerability - this branch is the replacement of d3 libraries with victory-vendor d3 libraries. D3 versions align with versions specified in https://github.com/FormidableLabs/victory/blob/main/packages/victory-vendor/package.json. This build is passing.

`d3-scale-chromatic` is only used in tests/demos so I upgraded it verbatim.

**Comment by jmfrancois** [comment]
This looks strange to me to go into the direction of a vendor packaged libraries from an external opensource project.

This article article already old https://devclass.com/2021/06/15/d3-7-0-goes-all-in-on-ecmascript-modules/ that explains why d3 has done the move to ESM.

Most of the tickets are about compatibility issue with tools which already support esm but need configuration.
Jest for example there is multiple way to configure it to work with esm. I gave an example PR here: https://github.com/recharts/recharts/issues/2915

Nextjs there is an example here: https://github.com/recharts/recharts/issues/2918

So why not just do the upgrade d3-* in major release ?

ESM is the future and tools should supports it more and more.

**Comment by ckifer** [comment]
People want a fix for the CVE, people want a patch that does not break their applications. 2.1.13/2.1.14 should've been released as 3.0 and we could've moved on from there. 

Since we find ourselves where we do now the choices are 3.0 with the breaking change or this. This provides a CVE fix for those on 2.x.x that might not ever be able to use ESM (I know they _should_ be able to, but this is the same reasoning as victory [in their blogpost](https://formidable.com/blog/2022/victory-esm/) - we certainly have consumers who "can't")

We can still do the breaking change in v3, revert back to d3 rather than `victory-vendor`, and drop support for Node apps using cjs to continue the trudge forward towards ESM support. For now this seems a decent solution to prevent more people from needing to migrate charting libraries due to a security risk and to fix pressing issue in 2.x.x. 

This might not be the best action that could be taken but it is action and that's better than none. If there is a lot of pushback against this change I'd be happy to help push for a 3.0 release instead.

**Comment by ckifer** [comment]
Change is released in alpha version `v2.3.0-alpha.1` (along with other small changes since 2.2)

https://www.npmjs.com/package/recharts/v/2.3.0-alpha.1

https://github.com/recharts/recharts/releases/tag/v2.3.0-alpha.1

Please help me test this so we can get 2.3.0 out in a week or so. Feel free to report here or tag me in issues with any bugs. Thanks all!

**Comment by mikegoatly** [comment]
@ckifer My use-case is fairly trivial, but it worked fine in my create-react-app based application

**Comment by nfiacco** [comment]
No problems here either! Just have simple bar and line charts though.

**Comment by wmhartl** [comment]
No problems here. Relatively straightforward responsive bar, line, and radar charts with default and custom labels, tooltips, and legends.

**Comment by esgem** [comment]
Apparently no problems here either. Simple graph with one or two bars per datapoint and according tooltips.

**Comment by firstclown** [comment]
I have a few charts:

- a fairly complex ComposedChart with Area, ReferenceAreas and an image based ReferenceDot with ChartLegend. Looks good to me. (and no errors or security warnings!) 👍
- a few horizontal BarCharts with ranges and Labels and XAxis and YAxis details 👍
- a section of Pie chart that we're using to show a gauge type visual 👍
- AreaChart with gradient Area underneath and tooltips 👍
- a Scatter with ReferenceLine with dots of different colors in ReferenceArea sections 👍

Looks great to me. I'd love to push this code live.

**Comment by ckifer** [comment]
Thanks for the responses all - we will most likely release sometime next week pending no bugs reported. Trying to take more precautions and do more testing than has been done in the past 🚀

**Comment by ckifer** [comment]
@ckotyan please try removing your `node_modules`, `package-lock.json`, and trying again. I can't reproduce that in an npm audit with `2.3.0-alpha.1`

**Comment by ckifer** [comment]
Resolved in `v2.3`+ - https://github.com/recharts/recharts/releases/tag/v2.3.0

Let me know if anyone experiences issues. Thanks!

**Comment by andresebr** [comment]
The vulnerability issue has been fixed as mentioned, but after upgrading to `v2.3.2`, every page where a LineChart should be displayed crashes (see screenshots for errors). No issues when using `v2.2.0`. 

![Screen Shot 2023-01-25 at 5 21 36 PM](https://user-images.githubusercontent.com/7132215/214705078-2128c080-4d99-4e7c-837d-138199c2615a.png)
![Screen Shot 2023-01-25 at 5 21 44 PM](https://user-images.githubusercontent.com/7132215/214705085-5e76e3c2-829e-42a3-a634-76f763a44fbd.png)

**Comment by ckifer** [comment]
Any details on your setup @andresebr 

Need a reproduction in order to address. Not failing anywhere we've tested.

Browser, framework if any, SSR or not, React version, TypeScript version

**Comment by andresebr** [comment]
> Any details on your setup @andresebr
> 
> Need a reproduction in order to address. Not failing anywhere we've tested.
> 
> Browser, framework if any, SSR or not, React version, TypeScript version

Edge `v107.0.1418.42`, no framework, no SSR, React `v18.2.0`, typescript `v4.9.4`. Also, just in case, here's how the component that handles Line Charts looks:

```js
export const LineChart: React.FunctionComponent<LineChartProps> = (
  props: LineChartProps
) => {
  return (
    <ResponsiveContainer width="90%" height={props.height}>
      <ReLineChart
        width={props.width}
        height={props.height}
        margin={{ top: 5, bottom: 5 }}
      >
        <CartesianGrid strokeDasharray="3 3" />
        <XAxis dataKey="x" type="category" allowDuplicatedCategory={false} />
        <YAxis dataKey="y" />
        <Tooltip />
        <Legend />
        {props.series.map((e, i) => (
          <Line
            dataKey="y"
            data={e.points}
            name={e.label}
            key={`Reference.${i}`}
            stroke={e.color}
          />
        ))}
        {props.references &&
          props.references.map((r) => (
            <ReferenceLine
              y={r.value}
              label={r.label}
              key={`Reference.${r.label}`}
              stroke={r.color}
              ifOverflow="extendDomain"
            />
          ))}
      </ReLineChart>
    </ResponsiveContainer>
  );
};
```

**Comment by ckifer** [comment]
@andresebr does it crash on other browsers?

**Comment by ckifer** [comment]
@andresebr https://codesandbox.io/s/recharts-crash-repro-144f85

I can't reproduce on CodeSandbox which makes me think a Browser issue with edge. Please confirm if your site works on other browsers. I can't test with edge.

**Comment by andresebr** [comment]
@ckifer Same issue on other browsers. Tested on Firefox and Chrome. The CodeSandbox example works fine everywhere, though. 

Now I'm wondering if it has something to do with my setup. It's a bit strange that there are no problems with `v2.2.0` except for the vulnerability issues. I will file a new issue and include all the details that might be relevant. Thanks!

**Comment by nikolasrieble** [comment]
@andresebr Could you please for the sake of completeness include a toy dataset with which your problem surfaces?

**Comment by nikolasrieble** [comment]
https://github.com/mbostock/internmap/issues/3

**Comment by nikolasrieble** [comment]
https://github.com/Microsoft/TypeScript/issues/10853
